### PR TITLE
[SPEC] v2.1 blink input types

### PR DESCRIPTION
**Pull request by nickfrosty**
## TLDR

Add new input types within blink clients to improve the user experience and unlock new use cases for Action builders.

## Rationale

The current Solana Actions and blinks specification only supports a single, generic input type of a non-structured text box. While this basic input is useful, having more explicit and declarative input types would allow Action builders and blink clients to unlock new use cases and improve the user experience of using blinks across the internet.

## SRFC

See SRFC #29 - https://forum.solana.com/t/srfc-29-input-types-of-blinks-and-actions/1804

The `next` tag has been published with this PR at version `2.1.0-beta`

```
npm install @solana/actions-spec@2.1.0-beta
```

### CODE REVIEW

**Comment by thearyanag** [comment]
lgtm @nickfrosty

File: packages/actions-spec/index.d.ts [hunk 1]
```diff
@@ -88,6 +94,35 @@ export interface ActionParameter {
   required?: boolean;
 }
 
+/**
+ * Input field type to present to the user. Normally resembling the respective
+ * [HTML `input` types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input)
+ * or standard HTML element (e.g. `select`) for the platform being used (web, mobile, etc).
+ * @default `text`
+ */
+export type ActionParameterType =
```
Comment (fsher) [inline packages/actions-spec/index.d.ts:103]: I'd like to suggest that we type this a bit more strictly:

```ts
export type GeneralParameterType =
  | 'text'
  | 'email'
  | 'url'
  | 'number'
  | 'date'
  | 'datetime-local'
  | 'textarea';

export type SelectableParameterType = 'select' | 'radio' | 'checkbox';

export type ParameterType = GeneralParameterType | SelectableParameterType;

export interface Parameter<T extends ParameterType> {
  /** input field type */
  type?: T;
  /** regular expression pattern to validate user input client side */
  pattern?: string;
  /** human-readable description of the `pattern` */
  patternDescription?: string;
  /** parameter name in url */
  name: string;
  /** input placeholder */
  label?: string;
  /** input required */
  required?: boolean;
}

export interface ParameterSelectable<T extends ParameterType>
  extends Parameter<T> {
  pattern?: never;
  patternDescription?: never;
  options: Array<{
    /** displayed UI label of this selectable option */
    label: string;
    /** value of this selectable option */
    value: string;
    /** whether this option should be selected by default */
    selected?: boolean;
  }>;
}

export type TypedParameter<T extends ParameterType = ParameterType> =
  T extends SelectableParameterType ? ParameterSelectable<T> : Parameter<T>;
```

that way, only `select`, `checkbox` and `radio` would have `options` and should never receive `pattern` and `patternDescription`

Can be verified like so:
```ts
const test: TypedParameter = {}; // ignore ts error here

if (test.type === 'select') {
  test.options; // no error
}

if (test.type === 'text') {
  test.options; // TS2339: Property options does not exist on type Parameter<"text">
}
```

Comment (zhelezkov) [reply packages/actions-spec/index.d.ts:103]: def agree with Filipp. More stricter types we have - more errors we can eliminate at build time, and avoid extra support questions

Comment (nickfrosty) [reply packages/actions-spec/index.d.ts:103]: Love this! A great addition and will help improve the DX of using the types

Comment (nickfrosty) [reply packages/actions-spec/index.d.ts:120]: I just pushed the updates and update the readme in a few places. I added a statement noting that the code blocks in the readme may be the simplified types and the actual ones in the `index.d.ts` may be more complicated for better type safety and DX

File: packages/actions-spec/README.md [hunk 2]
```diff
@@ -225,20 +235,160 @@ export interface LinkedAction {
   /** button text rendered to the user */
   label: string;
   /** Parameter to accept user input within an action */
-  parameters?: [ActionParameter];
+  parameters?: Array<ActionParameter | ActionParameterSelectable>;
```
Comment (fsher) [inline packages/actions-spec/README.md:238]: Given the d.ts file now uses a new type, should this be updated as well?

File: packages/actions-spec/index.d.ts [hunk 3]
```diff
@@ -73,27 +73,91 @@ export interface LinkedAction {
...
+  max?: M;
+}
+
+type MinMax<T extends ActionParameterType> = T extends "date" | "datetime-local"
+  ? string
+  : T extends "radio" | "select"
+  ? never
+  : number;
+
+type GeneralParameterType =
```
Comment (fsher) [inline packages/actions-spec/index.d.ts:113]: Let's make these exportable as well, so that it would be easy to narrow types (without redefining them)

File: packages/actions-spec/README.md (hunk 2 above)
Comment (nickfrosty) [reply packages/actions-spec/README.md:238]: I added a note to the top of this readme (and evnetually to the published docs) that the types/interfaces shown in here may be simplified version of the final.

do you think it is worth adding the `TypedActionParameter` definition into the readme?

File: packages/actions-spec/index.d.ts (hunk 3 above)
Comment (nickfrosty) [reply packages/actions-spec/index.d.ts:113]: Can do!

File: packages/actions-spec/README.md (hunk 2 above)
Comment (nickfrosty) [reply packages/actions-spec/README.md:238]: and since a dev could also still access these specific types based on the parent definition (like `ActionParameter<"email">["min"]`) they are still accessible. This way they will also get the correct type based on the parameter `type` too

Comment (fsher) [reply packages/actions-spec/README.md:238]: Ah, makes sense. Yeah, that's fine, just concerned that it's likely that a dev would copy-paste or look for certain types based on the doc. So the closer the better.

We can also do something like:

```ts
/**
 @see {TypedParameter}
*/
parameters?: Array<ActionParameter | ActionParameterSelectable>;
```

or the other way around:

```ts
/**
  @see {ActionParameter}
  @see {ActionParameterSelectable}
*/
parameters?: Array<TypedParameter>;
```

The closer to the actual types the better, but I agree - simplified is good

Comment (nickfrosty) [reply packages/actions-spec/README.md:238]: I like the `@see {TypedParameter}` idea. Let's go with that!

I could add a more explicit message too, something like:
```
/**
   @see `TypedParameter` for the non-simplified type
*/
parameters?: Array<ActionParameter | ActionParameterSelectable>;
```

That way we get the best of both worlds I think. Simplified in the readme, but being explicit for people when they read it.

I would be super surprised if people tried to copy pasta the type/interfaces out of the doc since there is not really a reason to do so. But I guess you never know

Comment (nickfrosty) [reply packages/actions-spec/README.md:238]: I like this option:

```
/**
  @see {ActionParameter}
  @see {ActionParameterSelectable}
*/
parameters?: Array<TypedParameter>;
```

That way we get the best of both worlds I think. Simplified in the readme, but being explicit for people when they read it.

I would be super surprised if people tried to copy pasta the type/interfaces out of the doc since there is not really a reason to do so. But I guess you never know

**Review by fsher — APPROVED** [review]
LGTM! Thank you @nickfrosty!
